File: src/main/java/com/viartemev/requestmapper/GoToRequestMappingAction.java
package com.viartemev.requestmapper;

import com.intellij.ide.actions.GotoActionBase;
import com.intellij.ide.util.gotoByName.ChooseByNamePopup;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;

import static com.intellij.openapi.actionSystem.CommonDataKeys.PROJECT;

public class GoToRequestMappingAction extends GotoActionBase implements DumbAware {

    @Override
    protected void gotoActionPerformed(AnActionEvent e) {
        Project project = e.getData(PROJECT);
        if (project == null) {
            return;
        }

        showNavigationPopup(e, new RequestMappingModel(project), new GoToRequestMappingActionCallback(), false);
    }

    private static final class GoToRequestMappingActionCallback extends GotoActionCallback<String> {
        @Override
        public void elementChosen(ChooseByNamePopup popup, Object element) {
            if (!(element instanceof RequestMappingItem)) {
                return;
            }

            RequestMappingItem item = (RequestMappingItem) element;
            if (item.canNavigate()) {
                item.navigate(true);
            }
        }
    }
}

File: src/main/kotlin/com/viartemev/requestmapper/GoToRequestMappingAction.kt
package com.viartemev.requestmapper

import com.intellij.ide.actions.GotoActionBase
import com.intellij.ide.util.gotoByName.ChooseByNamePopup
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys.PROJECT
import com.intellij.openapi.project.DumbAware
import com.viartemev.requestmapper.extensions.Extensions


class GoToRequestMappingAction : GotoActionBase(), DumbAware {

    override fun gotoActionPerformed(e: AnActionEvent) {
        val project = e.getData(PROJECT) ?: return
        val requestMappingModel = RequestMappingModel(project, Extensions.getExtensions())
        showNavigationPopup(e, requestMappingModel, GoToRequestMappingActionCallback(), null, true, false)
    }

    private class GoToRequestMappingActionCallback : GotoActionBase.GotoActionCallback<String>() {

        override fun elementChosen(popup: ChooseByNamePopup, element: Any) {
            if (element is RequestMappingItem && element.canNavigate()) {
                element.navigate(true)
            }
        }
    }
}
